Source/DafnyCore/Backends/Dafny/DafnyCodeGenerator.cs

```diff
@@ -177,39 +202,49 @@ class DafnyCodeGenerator : SinglePassCodeGenerator {
 
     protected override IClassWriter CreateTrait(string name, bool isExtern, List<TypeParameter> typeParameters,
       TraitDecl trait, List<Type> superClasses, IToken tok, ConcreteSyntaxTree wr) {
-      throw new UnsupportedFeatureException(Token.NoToken, Feature.Traits);
-      // traits need a bit more work
 
-      // if (currentBuilder is TraitContainer builder) {
-      //   List<DAST.Type> typeParams = new();
-      //   foreach (var tp in trait.TypeArgs) {
-      //     typeParams.Add((DAST.Type)DAST.Type.create_TypeArg(Sequence<Rune>.UnicodeFromString(IdProtect(tp.GetCompileName(Options)))));
-      //   }
+      if (currentBuilder is TraitContainer builder) {
+        List<DAST.Type> typeParams = new();
+        foreach (var tp in trait.TypeArgs) {
+          typeParams.Add((DAST.Type)DAST.Type.create_TypeArg(Sequence<Rune>.UnicodeFromString(IdProtect(tp.GetCompileName(Options)))));
+        }
 
-      //   return new ClassWriter(this, builder.Trait(name, typeParams));
-      // } else {
-      //   throw new InvalidOperationException();
-      // }
+        return new ClassWriter(this, typeParameters.Any(), builder.Trait(name, typeParams));
+      } else {
+        throw new InvalidOperationException();
+      }
     }
 
     protected override ConcreteSyntaxTree CreateIterator(IteratorDecl iter, ConcreteSyntaxTree wr) {
-      throw new UnsupportedFeatureException(Token.NoToken, Feature.Iterators);
+      ThrowSpecificUnsupported(iter.tok, Feature.Iterators);
+      return wr;
     }
 
-    private static bool isTpSupported(TypeParameter tp) {
-      return tp.Variance == TypeParameter.TPVariance.Non &&
-        tp.Characteristics.EqualitySupport == TypeParameter.EqualitySupportValue.Unspecified;
+    private static bool isTpSupported(TypeParameter tp, [CanBeNull] out string why) {
+      if (tp.Variance != TypeParameter.TPVariance.Non) {
+        why = $"<b>Unsupported: <i>Type variance {tp.Variance} not supported</i></b>";
+        return false;
+      }
+
+      if (tp.Characteristics.EqualitySupport != TypeParameter.EqualitySupportValue.Unspecified) {
+        why = $"<b>Unsupported: <i>Type parameter Equality support {tp.Characteristics.EqualitySupport} not supported for type parameters</i></b>";
+        return false;
+      }
+
+      why = null;
+      return true;
     }
 
     protected override IClassWriter DeclareDatatype(DatatypeDecl dt, ConcreteSyntaxTree wr) {
       if (currentBuilder is DatatypeContainer builder) {
         List<DAST.Type> typeParams = new();
         foreach (var tp in dt.TypeArgs) {
-          if (!isTpSupported(tp) && !(dt is TupleTypeDecl)) {
-            throwGeneralUnsupported(); //("Contravariance in type parameters");
+          var compileName = IdProtect(tp.GetCompileName(Options));
+          if (!isTpSupported(tp, out var why) && !(dt is TupleTypeDecl)) {
+            compileName += $"({why})";
```

Comment:
Isn't this a case that should call throwGeneralUnsupported()? Otherwise we might omit incorrect code without any warnings/errors.